src/components/scrollManager.js

```diff
@@ -0,0 +1,246 @@
+        if (delta1 < 0 && delta1 < delta2) {
+            return delta1;
+        } else if (delta2 < 0) {
+            return -delta2;
+        }
+        return 0;
+    };
+
+    function trend(v, v1, v2) {
+        return (v1 < v && v2 < v1) || (v1 > v && v2 > v1) ? v2 : v1;
+    };
+
+    function getWindowRect() {
+        return {
+            left: window.scrollX,
+            top: window.scrollY,
+            right: window.scrollX + window.innerWidth - 1,
+            bottom: window.scrollY + window.innerHeight - 1
+        };
+    }
+
+    function getViewRect() {
+        return {
+            left: window.scrollX + scrollExtra,
+            top: window.scrollY + scrollExtra,
+            right: window.scrollX + window.innerWidth - 1 - scrollExtra,
+            bottom: window.scrollY + window.innerHeight - 1 - scrollExtra
+        };
+    }
+
+    function getItemRect(elem) {
+        var rect = elem.getBoundingClientRect();
+        return {
+            left: rect.left + window.scrollX,
+            top: rect.top + window.scrollY,
+            right: rect.right + window.scrollX,
+            bottom: rect.bottom + window.scrollY
+        };
+    }
+
+    var origScrollTo = window.scrollTo;
+    var origScrollBy = window.scrollBy;
+
+    window.scroll = function(x, y) {
+        return scrollTo(x, y, false);
+    };
+
+    window.scrollTo = function(options) {
+        if (typeof options !== 'object') {
+            return scrollTo(arguments[0], arguments[1]);
+        }
+        return scrollTo(options.left, options.top, options.behavior === 'smooth');
+    };
+
+    window.scrollBy = function scrollTo(options) {
+        if (typeof options !== 'object') {
+            return scrollBy(arguments[0], arguments[1]);
+        }
+        return scrollBy(options.left, options.top, options.behavior === 'smooth');
+    };
+
+    function resetScroll() {
+        cancelAnimationFrame(scrollTimer);
+        scrollTimer = undefined;
+    }
+
+    function scrollTo(x, y, smooth) {
+        //console.log('scrollTo (' + x + ', ' + y + ', ' + smooth + ')');
+
+        if (smooth !== true) {
+            resetScroll();
+            origScrollTo(x, y);
+            return true;
+        }
+
+        if (scrollTimer) {
+            scrollTarget.x = trend(window.scrollX, scrollTarget.x, x);
+            scrollTarget.y = trend(window.scrollY, scrollTarget.y, y);
+        } else {
+            scrollTarget.x = x;
+            scrollTarget.y = y;
+        }
+
+        var scrollMaxX = Math.max(0, Math.max(document.body.scrollWidth, document.body.offsetWidth, document.documentElement.clientWidth, document.documentElement.scrollWidth, document.documentElement.offsetWidth) - window.innerWidth);
+        var scrollMaxY = Math.max(0, Math.max(document.body.scrollHeight, document.body.offsetHeight, document.documentElement.clientHeight, document.documentElement.scrollHeight, document.documentElement.offsetHeight) - window.innerHeight);
```

Comment:
please change those lines into multiline statements, otherwise reading them is painfully hard